### PR TITLE
[DHIS2-4890] The Analytics metadata response contains dynamic items info

### PR DESCRIPTION
**Pull request by luciano-fiandesio**
When e.g. LEVEL-2 is passed as an OrgUnit item to analytics,
		 the name of the Level (e.g. District) is also returned in the
		 metadata

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseDimensionalObject.java [hunk 1]
```diff
@@ -348,6 +362,17 @@ public void setDataDimension( boolean dataDimension )
         this.dataDimension = dataDimension;
     }
 
+    @Override
+    @JsonProperty
+    @JacksonXmlProperty( namespace = DxfNamespaces.DXF_2_0 ) // TODO do we need to serialize this field
+    public DimensionalAggregation getDimensionalAggregation() {
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseDimensionalObject.java]: Codestyle. Brackets on new line.

File: dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/DhisConvenienceTest.java [hunk 2]
```diff
@@ -1,2269 +1,2269 @@
...
- * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
- * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
- * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
- * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- */
-
-import com.google.common.collect.Lists;
-import com.google.common.collect.Sets;
-import com.google.common.hash.Hashing;
-import org.apache.commons.logging.Log;
```
Comment (larshelge) [inline dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/DhisConvenienceTest.java]: Could we avoid reformatting entire class?

File: dhis-2/pom.xml [hunk 3]
```diff
@@ -1,1405 +1,1404 @@
-<?xml version="1.0" encoding="UTF-8"?>
-
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (larshelge) [inline dhis-2/pom.xml]: Could we avoid reformatting entire class?

File: dhis-2/dhis-api/pom.xml [hunk 4]
```diff
@@ -142,6 +142,10 @@
       <groupId>com.vividsolutions</groupId>
       <artifactId>jts-io</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.projectlombok</groupId>
+      <artifactId>lombok</artifactId>
+    </dependency>
```
Comment (larshelge) [inline dhis-2/dhis-api/pom.xml]: I am wondering if the use of lombok is a system-wide decision - perhaps better not to use this in just a few places for consistency reasons?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalAggregation.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package org.hisp.dhis.common;
+
+import java.util.List;
+
+import lombok.Data;
+
+@Data
+public class DimensionalAggregation
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalAggregation.java]: I am wondering if *DimensionalKeywords* could be a more descriptive name - things like "level" is not really aggregating, rather selecting, items.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/AnalyticsUtils.java [hunk 6]
```diff
@@ -599,9 +599,15 @@ else if ( DimensionItemType.PROGRAM_INDICATOR == item.getDimensionItemType() &&
                         map.put( coc.getUid(), new MetadataItem( coc.getDisplayProperty( params.getDisplayProperty() ), includeMetadataDetails ? coc : null ) );
                     }
                 }
+
+
             }
 
             map.put( dimension.getDimension(), new MetadataItem( dimension.getDisplayProperty( params.getDisplayProperty() ), includeMetadataDetails ? dimension : null ) );
+            if (dimension.getDimensionalAggregation() != null) {
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/AnalyticsUtils.java]: Code style, brackets on new line.

File: dhis-2/dhis-api/pom.xml (hunk 4 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-api/pom.xml]: Yes, I wanted to discuss this with you before reviewing. We briefly discussed the use of such frameworks during the interview. I agree that it doesn't make a lot of sense to use Lombok or similar tools in a "spotty" way. It was more a way to start the discussion.

File: dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/DhisConvenienceTest.java (hunk 2 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/DhisConvenienceTest.java]: Yes, very much. I don't know exactly what happened here. I'm using the official DHIS2 coding formatting file for eclipse with an Intellij Plugin that uses the Eclipse engine for formatting. I need to investigate, sorry about that.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalAggregation.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package org.hisp.dhis.common;
+
+import java.util.List;
+
+public class DimensionalAggregation
+{
+    private final List<BaseIdentifiableObject> groupBy;
+
+    public DimensionalAggregation(List<BaseIdentifiableObject> groupBy) {
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalAggregation.java]: Code style, brackets.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/AnalyticsUtils.java [hunk 8]
```diff
@@ -599,9 +599,16 @@ else if ( DimensionItemType.PROGRAM_INDICATOR == item.getDimensionItemType() &&
                         map.put( coc.getUid(), new MetadataItem( coc.getDisplayProperty( params.getDisplayProperty() ), includeMetadataDetails ? coc : null ) );
                     }
                 }
+
+
             }
 
             map.put( dimension.getDimension(), new MetadataItem( dimension.getDisplayProperty( params.getDisplayProperty() ), includeMetadataDetails ? dimension : null ) );
+            if (dimension.getDimensionalAggregation() != null)
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/AnalyticsUtils.java]: Code style, space between ( )

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/AnalyticsServiceMetadataTest.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
+package org.hisp.dhis.analytics.data;
+
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/AnalyticsServiceMetadataTest.java]: Add license.

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/DefaultDataQueryServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
+package org.hisp.dhis.analytics.data;
+
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/DefaultDataQueryServiceTest.java]: Add license.

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/AnalyticsServiceMetadataTest.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+import static com.google.common.collect.Lists.newArrayList;
+import static org.hamcrest.CoreMatchers.*;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/AnalyticsServiceMetadataTest.java]: Add license.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalObjectUtils.java [hunk 12]
```diff
@@ -292,7 +280,7 @@ public static void setDimensionItemsForFilters( DimensionalObject dimension, Gri
 
         List<Object> values = new ArrayList<>( grid.getUniqueValues( dim.getDimension() ) );
 
-        Collections.sort( values, ObjectStringValueComparator.INSTANCE );
+        values.sort(ObjectStringValueComparator.INSTANCE);
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalObjectUtils.java]: Code style, space after ( )

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalObjectUtils.java [hunk 13]
```diff
@@ -419,9 +407,7 @@ public static boolean isWildCard( String identifier )
      */
     public static List<DimensionalItemObject> asList( Collection<? extends DimensionalItemObject> collection )
     {
-        List<DimensionalItemObject> list = new ArrayList<>();
-        list.addAll( collection );
-        return list;
+        return new ArrayList<>(collection);
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalObjectUtils.java]: Code style, space after ( )

File: dhis-2/dhis-api/src/test/java/org/hisp/dhis/common/BaseDimensionalObjectTest.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+                )
+        );
+        target.setFilter( "test-filter" );
+        target.setLegendSet(
+            new LegendSet( "legend-name", "symbolizer-test", Sets.newHashSet(
+                    buildLegend(),
+                    buildLegend()
+            ) ) );
+        target.setAggregationType(AggregationType.AVERAGE);
+        target.setDataDimension(true);
```
Comment (larshelge) [inline dhis-2/dhis-api/src/test/java/org/hisp/dhis/common/BaseDimensionalObjectTest.java]: Code style, space after ( )

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultDataQueryService.java [hunk 15]
```diff
@@ -454,15 +426,23 @@ else if ( !inputIdScheme.is( IdentifiableProperty.UID ) || CodeGenerator.isValid
...
             List<OrganisationUnit> ousList = asTypedList( ous );
-
+            DimensionalAggregation dimensionalAggregation = null;
             if ( !levels.isEmpty() )
             {
                 orgUnits.addAll( sort( organisationUnitService.getOrganisationUnitsAtLevels( levels, ousList ) ) );
+                dimensionalAggregation = new DimensionalAggregation(
+                    levels.stream().map( l -> organisationUnitService.getOrganisationUnitLevelByLevel( l ) )
+                                              .filter(Objects::nonNull)
+                                              .collect( Collectors.toList() ) );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultDataQueryService.java]: Code style, superfluous indendation

File: dhis-2/pom.xml [hunk 16]
```diff
@@ -1,1405 +1,1411 @@
-<?xml version="1.0" encoding="UTF-8"?>
-
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
-  <modelVersion>4.0.0</modelVersion>
-  <groupId>org.hisp.dhis</groupId>
-  <artifactId>dhis</artifactId>
-  <version>2.32-SNAPSHOT</version>
```
Comment (larshelge) [inline dhis-2/pom.xml]: Can we avoid reformatting this file

Comment (luciano-fiandesio) [reply dhis-2/pom.xml]: Yes, I found a way to fix this issue. Hopefully, this is the last time I reformat this file. It has to do with line ending ON THIS PARTICULAR file (other pom files are not affected). Sorry about that

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalAggregation.java (hunk 5 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalAggregation.java]: Renamed to DimensionalKeywords

**Pull request merged**